File: db/src/main/java/onl/ycode/stormify/TableInfo.java
// SPDX-License-Identifier: Apache-2.0
// (C) Panayotis Katsaloulis

package onl.ycode.stormify;

import onl.ycode.stormify.FieldInfo.FieldContext;

import java.util.*;

import static onl.ycode.stormify.StormifyManager.stormify;
import static onl.ycode.stormify.Utils.*;

/**
 * Parse a class and handle it as a database table.
 */
public class TableInfo {
    private final Class<?> classType;
    private final String tableName;
    private final List<FieldInfo> fields;
    private final List<FieldInfo> primaryKeys;

    // Lazy cached data
    private final LazyProperty<Map<String, Collection<FieldInfo>>> dbFields;
    private final LazyProperty<Map<String, FieldInfo>> objFields;
    private final LazyProperty<List<FieldInfo>> createFields;
    private final LazyProperty<List<FieldInfo>> updateFields;
    final LazyProperty<String> createFieldNames;
    final LazyProperty<String> updateFieldNames;
    final LazyProperty<String> createPlaceholders;

    TableInfo(Class<?> classType, String tableName, Collection<FieldInfo> fields) {
        this.classType = classType;
        this.tableName = tableName;
        this.fields = fields instanceof List ? (List<FieldInfo>) fields : new ArrayList<>(fields);
        this.primaryKeys = filter(fields, FieldInfo::isPrimaryKey);
        this.dbFields = new LazyProperty<>(() -> {
            Map<String, Collection<FieldInfo>> result = new LinkedHashMap<>();
            for (FieldInfo field : fields)
                result.computeIfAbsent(field.getDbName(), k -> new ArrayList<>()).add(field);
            return result;
        });
        this.objFields = new LazyProperty<>(() -> {
            Map<String, FieldInfo> result = new LinkedHashMap<>();
            for (FieldInfo field : fields)
                result.put(field.getName(), field);
            return result;
        });
        this.createFields = new LazyProperty<>(() -> filter(fields, FieldInfo::isInsertable));
        this.updateFields = new LazyProperty<>(() -> filter(fields, FieldInfo::isUpdatable));
        this.createFieldNames = new LazyProperty<>(() -> String.join(", ", map(createFields.get(), FieldInfo::getDbName)));
        this.updateFieldNames = new LazyProperty<>(() -> String.join(", ", map(updateFields.get(), fieldInfo -> fieldInfo.getDbName() + " = ?")));
        this.createPlaceholders = new LazyProperty<>(() -> nCopies("?", ", ", updateFields.get().size()));
    }

    /**
     * Get the class type of the table.
     *
     * @return The class type of the table.
     */
    public Class<?> getClassType() {
        return classType;
    }

    /**
     * Get the name of the table.
     *
     * @return The name of the table.
     */
    public String getTableName() {
        return tableName;
    }

    /**
     * Get a list of all primary keys of the table.
     *
     * @return The primary keys of the table.
     */
    public List<FieldInfo> getPrimaryKeys() {
        return primaryKeys;
    }

    /**
     * Get the primary key of the table. If the table has more than one primary key, or
     * no primary key, then an exception is thrown.
     *
     * @return The primary key of the table.
     */
    public FieldInfo getPrimaryKey() {
        if (primaryKeys.size() == 1)
            return primaryKeys.get(0);
        else
            throw new QueryException("Table " + tableName + " needs to have exactly one primary key");
    }

    /**
     * Get a reference of the fields of the table.
     *
     * @return The fields of the table.
     */
    public Collection<FieldInfo> getFields() {
        return getFields(null);
    }

    List<FieldInfo> getFields(FieldContext ctx) {
        if (ctx == FieldContext.CREATE)
            return createFields.get();
        else if (ctx == FieldContext.UPDATE)
            return updateFields.get();
        else
            return fields;
    }

    /**
     * Get a reference of the field, given a specific database column name.
     * <p>
     * Note that there may be more than one FieldInfo, to support cases where the same column is used in multiple
     * fields. One such common scenario is when a foreign key is used both as a reference and as a numeric id value.
     *
     * @param name The name of the column in the database.
     * @return The fields corresponding to the specific column name.
     */
    public Collection<FieldInfo> getDbField(String name) {
        return dbFields.get().getOrDefault(name, Collections.emptyList());
    }

    /**
     * Get a reference of the field, given a specific field name.
     *
     * @param name The name of the field, as a Java property.
     * @return The field corresponding to the specific field name.
     */
    public FieldInfo getField(String name) {
        return objFields.get().get(name);
    }

    @Override
    public String toString() {
        return tableName + "{" +
                String.join(" ", map(fields, it -> "\uD83C\uDFF7" + it.toString())) +
                (primaryKeys.isEmpty() ? "" : " " + map(primaryKeys, it -> "\uD83C\uDFF7" + it.toString())) +
                '}';
    }

    /**
     * Check the consistency of the table. For every database column, there should be at most one field that is
     * updatable and at most one field that is creatable. If there are more than one fields, then the
     * table is considered inconsistent.
     *
     * @return True if the table is consistent, false otherwise.
     */
    public boolean checkConsistency() {
        Map<String, Collection<FieldInfo>> fieldConsistency = new LinkedHashMap<>();
        boolean foundError = false;
        for (FieldInfo field : fields)
            fieldConsistency.computeIfAbsent(field.getDbName(), k -> new ArrayList<>()).add(field);
        for (Map.Entry<String, Collection<FieldInfo>> entry : fieldConsistency.entrySet()) {
            Collection<FieldInfo> commonFields = entry.getValue();
            if (commonFields.size() > 1) {
                Collection<FieldInfo> createFields = filter(commonFields, FieldInfo::isInsertable);
                Collection<FieldInfo> updateFields = filter(commonFields, FieldInfo::isUpdatable);
                if (createFields.size() > 1)
                    stormify().getLogger().error("Table " + tableName + " has more than one field that is creatable for column " + entry.getKey() + ": " + createFields);
                if (updateFields.size() > 1)
                    stormify().getLogger().error("Table " + tableName + " has more than one field that is updatable for column " + entry.getKey() + " : " + updateFields);
                foundError |= createFields.size() > 1 || updateFields.size() > 1;
            }
        }
        if (foundError) {
            stormify().getLogger().error("Please use the appropriate annotations to mark fields as non-creatable or non-updatable.");
            stormify().getLogger().error("The relevant annotations are @" + DbField.class.getName() + ",  @javax.persistence.Column, and @javax.persistence.JoinColumn");
        }
        return !foundError;
    }
}

File: stormify/src/commonMain/kotlin/onl/ycode/stormify/TableInfo.kt
// SPDX-License-Identifier: Apache-2.0
// (C) Panayotis Katsaloulis
package onl.ycode.stormify

import onl.ycode.kdbc.SQLException
import onl.ycode.kdbc.TypeConversion
import kotlin.reflect.KClass

/**
 * Describes a single mapped field (property) of an entity, including its database column name,
 * type, and role in CRUD operations.
 *
 * @property name the Kotlin property name
 * @property dbName the corresponding database column name (after [NamingPolicy] conversion)
 * @property type the Kotlin type of the property
 * @property isPrimaryKey whether this field is part of the entity's primary key
 * @property isReference whether this field is a foreign key reference to another entity
 * @property isEnum whether the property type is a Kotlin/Java enum
 * @property enumAsString whether enum values are stored as their string name (`true`) or ordinal/custom integer (`false`)
 * @property sequence the database sequence name used to generate values, or null if not sequence-backed
 * @property isAutoIncrement whether the database auto-generates values for this field (e.g. IDENTITY columns)
 * @property isInsertable whether this field is included in INSERT statements
 * @property isUpdatable whether this field is included in UPDATE statements
 */
data class FieldInfo(
    val name: String,
    val dbName: String,
    val type: KClass<*>,
    val isPrimaryKey: Boolean,
    val isReference: Boolean,
    val isEnum: Boolean,
    val enumAsString: Boolean,
    val sequence: String?,
    val isAutoIncrement: Boolean,
    val isInsertable: Boolean,
    val isUpdatable: Boolean,
)

/**
 * Metadata container for a mapped entity class. Holds the table name, field mappings, primary key
 * information, and pre-built SQL queries for CRUD operations.
 *
 * Instances are created internally by [Stormify] and cached per entity class. Use
 * [Stormify.getTableInfo] to obtain the metadata for a given class.
 *
 * @property tableName the database table name this entity maps to
 */
class TableInfo<T : Any> internal constructor(
    private val meta: EntityMeta<T>,
    private val resolved: List<ResolvedProperty<T>>,
    val tableName: String,
) {
    /** The Kotlin class this metadata describes. */
    val classType: KClass<T> get() = meta.type

    internal fun create(): T = meta.constructor()

    // Resolved property lists
    private val idProps = resolved.filter { it.isPrimaryKey }
    private val restProps = resolved.filter { !it.isPrimaryKey }
    private val insertableProps = resolved.filter { it.isInsertable }
    private val updatableProps = resolved.filter { it.isUpdatable && !it.isPrimaryKey }

    // DB name lookups
    private val fieldTypeMap = resolved.associate { it.dbName.lowercase() to it.type }
    private val fieldByDbName = resolved.groupBy { it.dbName.lowercase() }
    private val referenceFieldMap = resolved.filter { it.isReference }.associate { it.dbName.lowercase() to it.type }
    private val enumFieldSet = resolved.filter { it.isEnum }.mapTo(HashSet()) { it.dbName.lowercase() }

    internal fun getType(dbName: String): KClass<*> =
        fieldTypeMap[dbName.lowercase()] ?: Any::class

    internal fun getScalarType(dbName: String): KClass<*>? =
        fieldTypeMap[dbName.lowercase()]?.takeIf { TypeConversion.isKnownScalar(it) }

    internal fun isReferenceField(dbName: String): Boolean =
        referenceFieldMap.containsKey(dbName.lowercase())

    internal fun getReferenceType(dbName: String): KClass<*>? =
        referenceFieldMap[dbName.lowercase()]

    internal fun isEnumField(dbName: String): Boolean =
        enumFieldSet.contains(dbName.lowercase())

    internal fun setField(
        entity: T,
        dbName: String,
        value: Any?,
        stormify: Stormify,
        policy: UnmatchedColumnPolicy = UnmatchedColumnPolicy.THROW,
    ) {
        val props = fieldByDbName[dbName.lowercase()]
        if (props.isNullOrEmpty()) {
            when (policy) {
                UnmatchedColumnPolicy.THROW -> throw SQLException(
                    "Column $dbName has no matching field in ${meta.type.simpleName}"
                )
                UnmatchedColumnPolicy.WARN -> stormify.logger.warn(
                    "Column $dbName has no matching field in ${meta.type.simpleName}"
                )
                UnmatchedColumnPolicy.IGNORE -> Unit
            }
            return
        }
        for (prop in props) prop.setter(entity, value, stormify)
    }

    /**
     * Returns the resolved properties for the column [dbName], pre-keyed by the lowercased
     * label. Empty list = no matching field on this entity (caller decides the
     * unmatched-column policy). Used by [Stormify]'s populate hot path to cache the property
     * dispatch per-ResultSet so that the per-row inner loop avoids the `dbName.lowercase()`
     * allocation and `fieldByDbName` HashMap lookup on every cell.
     */
    internal fun propertiesForColumn(dbName: String): List<ResolvedProperty<T>> =
        fieldByDbName[dbName.lowercase()] ?: emptyList()

    // ID operations (cached lists — avoid allocation on every call)
    internal val idDbNames = idProps.map { it.dbName }
    internal val idTypes = idProps.map { it.type }
    internal val idSequences = idProps.map { it.sequence ?: "" }

    internal val singleKeyDbName: String by lazy {
        if (idProps.size == 1) idProps[0].dbName
        else throw SQLException("Expected exactly one primary key in $tableName, found ${idProps.size}")
    }

    internal fun getIdValues(entity: T): List<Any?> = idProps.map { it.getter(entity) }

    // SELECT * is intentional; unmatched DB columns go through Stormify.unmatchedColumnPolicy.
    internal val populateQuery by lazy {
        "SELECT * FROM $tableName WHERE ${idProps.joinToString(" AND ") { "${it.dbName} = ?" }}"
    }
    internal val createQuery by lazy {
        val fields = insertableProps.joinToString(", ") { it.dbName }
        val placeholders = insertableProps.joinToString(", ") { "?" }
        "INSERT INTO $tableName ($fields) VALUES ($placeholders)"
    }
    internal val updateQuery by lazy {
        if (updatableProps.isEmpty())
            throw SQLException("Entity $tableName has no updatable fields")
        val setClause = updatableProps.joinToString(", ") { "${it.dbName} = ?" }
        val whereClause = idProps.joinToString(" AND ") { "${it.dbName} = ?" }
        "UPDATE $tableName SET $setClause WHERE $whereClause"
    }
    // Values for create/update queries (in query parameter order)
    internal fun getCreateValues(entity: T): List<Any?> = insertableProps.map { it.sqlValue(entity) }
    internal fun getUpdateValues(entity: T): List<Any?> =
        updatableProps.map { it.sqlValue(entity) } + getIdValues(entity)

    /** All mapped fields of this entity, including primary keys and regular columns. */
    val fieldInfos: List<FieldInfo> by lazy {
        resolved.map {
            FieldInfo(it.name, it.dbName, it.type, it.isPrimaryKey, it.isReference, it.isEnum, it.enumAsString,
                it.sequence, it.isAutoIncrement, it.isInsertable, it.isUpdatable)
        }
    }

    /** The primary key fields of this entity. May contain multiple entries for composite keys. */
    val primaryKeys: List<FieldInfo> get() = fieldInfos.filter { it.isPrimaryKey }

    /** The single primary key field. Throws if the entity has zero or more than one primary key. */
    val primaryKey: FieldInfo
        get() = primaryKeys.singleOrNull()
            ?: throw SQLException("Expected exactly one primary key in $tableName, found ${primaryKeys.size}")

    /** Finds a field by its Kotlin property [name] (case-insensitive), or null if not found. */
    fun getField(name: String): FieldInfo? =
        fieldInfos.find { it.name.equals(name, ignoreCase = true) }

    internal companion object {
        @Suppress("UNCHECKED_CAST")
        internal fun <T : Any> build(
            meta: EntityMeta<T>,
            namingPolicy: (String) -> String,
            blacklist: Set<String>,
            pkResolvers: Collection<(String, String) -> Boolean>
        ): TableInfo<T> {
            val tableName = meta.tableNameOverride?.takeIf { it.isNotBlank() }
                ?: namingPolicy(meta.type.simpleName ?: meta.type.toString())

            val activeProps = meta.properties
                .filter { !it.isTransient && it.name !in blacklist }

            val hasPkAnnotation = activeProps.any { it.isPrimaryKey }

            val resolved = activeProps.map { prop ->
                val dbName = prop.dbNameOverride?.takeIf { it.isNotBlank() }
                    ?: namingPolicy(prop.name)
                val isPk = if (hasPkAnnotation) prop.isPrimaryKey
                else pkResolvers.any { resolver -> resolver(tableName, prop.name) }
                ResolvedProperty(
                    name = prop.name, dbName = dbName, type = prop.type,
                    isReference = prop.isReference, isEnum = prop.isEnum,
                    enumAsString = prop.enumAsString,
                    isPrimaryKey = isPk,
                    sequence = prop.sequence, isAutoIncrement = prop.isAutoIncrement,
                    isInsertable = prop.isCreatable,
                    isUpdatable = prop.isUpdatable,
                    getter = prop.getter, setter = prop.setter
                )
            }

            return TableInfo(meta, resolved, tableName)
        }
    }
}

internal class ResolvedProperty<T : Any>(
    val name: String,
    val dbName: String,
    val type: KClass<*>,
    val isReference: Boolean,
    val isEnum: Boolean,
    val enumAsString: Boolean,
    val isPrimaryKey: Boolean,
    val sequence: String?,
    val isAutoIncrement: Boolean,
    isInsertable: Boolean,
    val isUpdatable: Boolean,
    val getter: (T) -> Any?,
    val setter: (T, Any?, Stormify) -> Unit,
) {
    val isInsertable: Boolean = isInsertable && !isAutoIncrement

    fun sqlValue(entity: T): Any? {
        val value = getter(entity)
        if (value is Enum<*>) return if (enumAsString) value.name else enumToInt(value)
        return value
    }
}
